js/integration-tests/solidity-tests/test/fuzz.ts

```diff
@@ -0,0 +1,105 @@
+  });
+
+  it("FailingFuzz", async function () {
+    const failureDir = testContext.fuzzFailuresPersistDir;
+
+    // Remove invariant config failure directory to make sure it's created fresh.
+    await fs.rm(failureDir, {
+      recursive: true,
+      force: true,
+    });
+
+    const result1 = await testContext.runTestsWithStats("OverflowTest");
+    assert.equal(result1.failedTests, 1);
+    assert.equal(result1.totalTests, 1);
+
+    // The fuzz failure directory should not be created if we don't set the directory
+    assert.isFalse(existsSync(failureDir));
+
+    const result2 = await testContext.runTestsWithStats("OverflowTest", {
+      fuzz: {
+        failurePersistDir: failureDir,
+      },
+    });
+    assert.equal(result2.failedTests, 1);
+    assert.equal(result2.totalTests, 1);
+
+    // The fuzz failure directory should now be created
+    assert.isTrue(existsSync(failureDir));
+  });
+
+  // One test as steps should be sequential
+  it("FailingInvariant", async function () {
+    const failureDir = testContext.invariantFailuresPersistDir;
+    const defaultConfig = {
+      depth: 15,
+      // This is false by default, we just specify it here to make it obvious to the reader.
+      failOnRevert: false,
+    };
+
+    // Remove invariant config failure directory to make sure it's created fresh.
+    await fs.rm(failureDir, {
+      recursive: true,
+      force: true,
+    });
+
+    const result1 = await testContext.runTestsWithStats(
+      "FailingInvariantTest",
+      {
+        invariant: defaultConfig,
+      }
+    );
+    assert.equal(result1.failedTests, 1);
+    assert.equal(result1.totalTests, 1);
+
+    // The invariant failure directory should not be created if we don't set the directory
+    assert.isFalse(existsSync(failureDir));
+
+    const firstStart = performance.now();
+    const result2 = await testContext.runTestsWithStats(
+      "FailingInvariantTest",
+      {
+        invariant: {
+          ...defaultConfig,
+          failurePersistDir: failureDir,
+        },
+      }
+    );
+    const firstDuration = performance.now() - firstStart;
+    assert.equal(result2.failedTests, 1);
+    assert.equal(result2.totalTests, 1);
+
+    // The invariant failure directory should now be created
+    assert.isTrue(existsSync(failureDir));
+
+    const secondStart = performance.now();
+    const result3 = await testContext.runTestsWithStats(
+      "FailingInvariantTest",
+      {
+        invariant: {
+          ...defaultConfig,
+          failurePersistDir: failureDir,
+        },
+      }
+    );
+    assert.equal(result2.failedTests, 1);
+    assert.equal(result2.totalTests, 1);
+    const secondDuration = performance.now() - secondStart;
+
+    // The invariant failure directory should still be there
+    assert.isTrue(existsSync(failureDir));
+    // The second run should be faster than the first because it should use the persisted failure.
+    assert.isBelow(secondDuration, firstDuration * 0.25);
```

Comment:
I think there is a more robust way to check this. The `TestResult` objects have a `kind.runs` number for invariant tests which will be 1 when there is a cached seed, and it will almost surely be above 1 when there's not (and this can be done even more unlikely by changing the `% 13` part of the solidity test to a bigger value).

The problem is that `runTestsWithStats` only returns the number of tests and failures. Maybe we can just add a `testResults` property to its returned value, that way there's no need to refactor all the existing usages of that helper.

This is just a suggestion btw, feel free to keep it as is if you think that this approach won't have frequent false positives or negatives.

Reply:
Thanks, good idea, switched to checking runs. The assertion is that there are more than 1 runs on an uncached run and there is 1 run on a cached one. I didn't want to increase the `13` to a larger value to reduce the chance of the invariant testing not catching the bug.

[ab94d0b](https://github.com/NomicFoundation/edr/pull/658/commits/ab94d0b77756dbc8b1caa6fdadb7db918b2fe2f6)